transport/http2_server.go

```diff
@@ -206,21 +202,13 @@ func (t *http2Server) operateHeaders(hDec *hpackDecoder, s *Stream, frame header
 		recv: s.buf,
 	}
 	s.method = hDec.state.method
-	if t.tracing {
-		s.tr = trace.New("grpc.Recv."+MethodFamily(s.method), s.method)
-		s.ctx = trace.NewContext(s.ctx, s.tr)
-	}
-	wg.Add(1)
-	go func() {
-		handle(s)
-		wg.Done()
-	}()
+	handle(s, wg)
 	return nil
 }
 
 // HandleStreams receives incoming streams using the given handler. This is
 // typically run in a separate goroutine.
-func (t *http2Server) HandleStreams(handle func(*Stream)) {
+func (t *http2Server) HandleStreams(handle func(*Stream, *sync.WaitGroup)) {
```

Comment:
It's usually a red flag for sync primitives to be passed around in APIs. I think it's a mistake for sync.WaitGroup args to be here and in operateHeaders. In general, it is a better idea for the caller to pass in a closure over the wait group, rather than squeeze it in here explicitly, particularly since there's no clear rules stated here about what's meant to be done with it.